solidity/ecdsa/README.adoc

```diff
@@ -15,7 +17,372 @@ endif::[]
+  signed version of a given transaction to be performed on behalf of the signing
+  group.
+
+== Prior Work
+
+Smart contract for the first version of Keep ECDSA ara available in
+link:https://github.com/keep-network/keep-ecdsa/tree/main/solidity[`keep-ecdsa` repository].
+The new version is optimised for larger groups by implementing optimistic
+selection of group members during DKG protocol. Staker rewards are redesigned
+and allocated for all sortition pool members. Most parameters are now governable.
+
+== The Mechanism
+
+=== Wallet Creation
+
+New wallet is created on request from wallet owner address. Signing group
+creation starts with an owner's call to `WalletRegistry.requestNewWallet()`.
+This transaction locks the sortition pool and sends a request to the Random
+Beacon for a new relay entry. From this moment, no operator can enter
+or leave the pool. Once a new relay entry appears on the chain, all off-chain
+clients perform group selection by calling `WalletRegistry.selectGroup()` view
+function for free. Relay entry provided by the Random Beacon is used as a seed
+for the group selection. After determining signing group members, clients should 
+perform off-chain distributed key generation (DKG).
+<<operator-only,One of the group members>> submits the result to the chain
+calling `WalletRegistry.submitDkgResult(DKG.Result calldata dkgResult)`
+function. Once the result is submitted, a challenge period starts.
+
+During the challenge period, anyone can notify that the submitted DKG result is
+malicious by calling `WalletRegistry.challengeDkgResult(DKG.Result calldata dkgResult)`
+function. A malicious DKG result may contain corrupted data, group members not
+selected by the pool, or incorrect supporting signatures. If such malicious
+result is submitted and successfully challenged, the result submitter gets
+slashed and the malicious result is immediately discarded. The address which
+notified about malicious DKG result is <<punishment,rewarded>>. DKG timeout
+timer is reset, and group members have another chance to submit a valid result.
+
+Once the challenge period passes, and no valid challenge is reported, the DKG
+result submitter should unlock the sortition pool and mark the DKG result as
+approved calling `WalletRegistry.approveDkgResult(DKG.Result calldata dkgResult)`.
+The submitter receives an ETH reimbursement for both `submitDkgResult` and
+`approveDkgResult` transaction as described in
+<<transaction-incentives,Transaction Incentives>> section. In case the original
+submitter does not call the `approveDkgResult` function within a specific number
+of blocks, anyone can do that and receive the submitter's reimbursement.
+
+There is a timeout before which a DKG result should be submitted.
+In case the DKG result was not submitted before a timeout, anyone can notify DKG
+timed out by calling `WalletRegistry.notifyDkgTimeout()` and unlock the sortition
```

Comment:
Maybe we can rephrase it a bit:
```
In case the DKG result was not submitted before the timeout, anyone can 
notify about the timed out DKG by calling `WalletRegistry.notifyDkgTimeout()`
function and unlock the sortition pool as part of that call.
```